File: DD_python/Examples/SequencingInstance/SequencingPathSolver.py
import time
from SourceCode.GraphAlgorithms.ShortestLongestPath.PathStructure import PathStructureSolution


class _NodeDP:
    """Best cost reaching a node plus the range its paths span, which prices the
    outgoing arcs: possible last jobs (bitmask with the +1 offset of `setup_times`),
    weight already scheduled (hi/lo) and jobs seen. One state per node -> a point."""

    __slots__ = ("cost", "previous_jobs", "scheduled_weight_hi",
                 "scheduled_weight_lo", "seen_jobs")

    def __init__(self, cost, previous_jobs, scheduled_weight_hi,
                 scheduled_weight_lo, seen_jobs):
        self.cost = cost
        self.previous_jobs = previous_jobs
        self.scheduled_weight_hi = scheduled_weight_hi
        self.scheduled_weight_lo = scheduled_weight_lo
        self.seen_jobs = seen_jobs


class SequencingPathSolver:
    """Shortest/longest path solver for Sequencing, drop-in for ShortestLongestPath.
    Σ w_j C_j is charged per arc as (setup + p[job]) * (W_total - already scheduled);
    both terms depend on the path, so each node covers all of its paths (`_NodeDP`)
    and its arcs take the extreme that keeps the bound (relaxed under, the rest over).
    """

    def __init__(self, dd: 'DD'):
        self._graph = dd.get_decision_diagram()
        self._problem = dd.problem
        self._relaxed: bool = dd.get_dd_kind() == 'relaxed'
        self._time: float = 0
        self._solution: PathStructureSolution = PathStructureSolution()
        self._objective: str = ""

    def set_parameters(self, weights: list, objective: str) -> None:
        self._objective = objective

    def get_time(self) -> str:
        return str(self._time)

    def get_solution(self) -> PathStructureSolution:
        return self._solution

    def solve(self) -> PathStructureSolution:
        print("\nStarting the process to solve the shortest or longest path in the graph...")
        start = time.time()
        self._run_dp()
        self._time = time.time() - start
        print("Shortest/longest path algorithm completed.")
        return self._solution

    def _run_dp(self) -> None:
        cost_sign = -1 if self._objective == "max" else 1
        dp_data, dp_arcs = self._initialize_dp_tables()
        self._forward_pass(dp_data, dp_arcs, cost_sign)
        self._extract_solution(dp_data, dp_arcs)

    def _initialize_dp_tables(self):
        total_nodes = self._graph.get_node_count()
        dp_data: list = [None] * total_nodes
        dp_arcs = [None] * total_nodes
        root_node = self._graph.structure[0][0]
        # Nothing is scheduled yet and the only possible previous job is the depot.
        dp_data[root_node.get_id()] = _NodeDP(0, 1 << 0, 0, 0, 0)
        return dp_data, dp_arcs

    def _best_setup(self, previous_jobs: int, job_id: int, charge_most: bool) -> int:
        """Setup for `job_id` over the possible previous jobs: the dearest one when
        charging the most, the cheapest one otherwise."""
        setup_times = self._problem.setup_times
        best = None
        remaining = previous_jobs
        while remaining:
            lowest_bit = remaining & -remaining
            candidate = setup_times[lowest_bit.bit_length() - 1][job_id]
            if best is None:
                best = candidate
            elif charge_most:
                best = max(best, candidate)
            else:
                best = min(best, candidate)
            remaining ^= lowest_bit
        return best

    def _describe_node(self, incoming, total_weight) -> _NodeDP:
        """Merge into one description everything the incoming paths can look like."""
        node_dp = _NodeDP(float("inf"), 0, 0, float("inf"), 0)
        for arc, parent in incoming:
            job_id = arc.variable_value
            job_bit = 1 << job_id
            job_weight = self._problem.weights[job_id]
            node_dp.previous_jobs |= 1 << (job_id + 1)
            node_dp.seen_jobs |= parent.seen_jobs | job_bit
            # A job already seen adds no weight, and the total is capped at W_total.
            node_dp.scheduled_weight_hi = max(
                node_dp.scheduled_weight_hi,
                min(total_weight, parent.scheduled_weight_hi + job_weight))
            node_dp.scheduled_weight_lo = min(
                node_dp.scheduled_weight_lo,
                parent.scheduled_weight_lo
                + (0 if parent.seen_jobs & job_bit else job_weight))
        return node_dp

    def _forward_pass(self, dp_data, dp_arcs, cost_sign) -> None:
        total_weight = sum(self._problem.weights)
        maximising = cost_sign < 0
        # Relaxed may not overcharge, the rest may not undercharge; max swaps it.
        charge_most = (self._relaxed == maximising)
        for layer in self._graph.structure[1:]:
            for node in layer:
                node_id = node.get_id()
                incoming = [(arc, dp_data[arc.parent_node.get_id()])
                            for arc in node.in_arcs
                            if dp_data[arc.parent_node.get_id()] is not None]
                if not incoming:
                    continue

                node_dp = self._describe_node(incoming, total_weight)

                for arc, parent in incoming:
                    job_id = arc.variable_value
                    setup = self._best_setup(parent.previous_jobs, job_id, charge_most)
                    duration = setup + self._problem.processing_times[job_id]
                    # duration * (W_total - already_scheduled), taking the end of
                    # the range that keeps the bound.
                    scheduled = parent.scheduled_weight_lo if charge_most \
                        else parent.scheduled_weight_hi
                    arc_cost = duration * (total_weight - scheduled)
                    candidate_cost = parent.cost + cost_sign * arc_cost
                    if candidate_cost < node_dp.cost:
                        node_dp.cost = candidate_cost
                        dp_arcs[node_id] = arc

                dp_data[node_id] = node_dp

    def _extract_solution(self, dp_data, dp_arcs) -> None:
        terminal_id = self._graph.structure[-1][0].get_id()
        optimal_cost = dp_data[terminal_id].cost
        if self._objective == "max":
            optimal_cost = -optimal_cost
        self._solution.value = optimal_cost
        self._solution.path_arcs, self._solution.path_print = self._backtrack_optimal_path(dp_arcs, terminal_id)

    def _backtrack_optimal_path(self, dp_arcs, terminal_id):
        path_arcs = []
        path_string = ""
        current_id = terminal_id
        while dp_arcs[current_id] is not None:
            arc = dp_arcs[current_id]
            path_string = f"-> {arc.to_string()}" + path_string
            path_arcs.insert(0, arc)
            current_id = arc.parent_node.get_id()
        return path_arcs, path_string[2:]

File: DD_c_plusplus/Examples/SequencingInstance/SequencingPathSolver.cpp
#include "SequencingPathSolver.h"
#include <iostream>
#include <algorithm>
#include <sstream>

SequencingPathSolver::SequencingPathSolver(DD<SequencingState>& dd)
    : graph_(dd.get_decision_diagram()),
      problem_(dynamic_cast<const SequencingProblem*>(&dd.problem)),
      relaxed_(dd.get_dd_kind() == DDKind::Relaxed),
      time_(chrono::duration<double>(0))
{}

void SequencingPathSolver::set_parameters(const vector<double>& /*weights*/, const string& objective) {
    objective_ = objective;
}

PathStructureSolution<SequencingState> SequencingPathSolver::solve() {
    cout << "\nStarting the process to solve the shortest or longest path in the graph...\n";
    auto start = chrono::steady_clock::now();
    run_dp();
    time_ = chrono::steady_clock::now() - start;
    cout << "Shortest/longest path algorithm completed.\n";
    return solution_;
}

PathStructureSolution<SequencingState> SequencingPathSolver::get_solution() const {
    return solution_;
}

string SequencingPathSolver::get_time() const {
    return to_string(time_.count());
}

void SequencingPathSolver::run_dp() {
    double cost_sign = (objective_ == "max") ? -1.0 : 1.0;
    vector<NodeDP> dp_data;
    vector<Arc<SequencingState>*> dp_arcs;
    initialize_dp_tables(dp_data, dp_arcs);
    forward_pass(dp_data, dp_arcs, cost_sign);
    extract_solution(dp_data, dp_arcs);
}

void SequencingPathSolver::initialize_dp_tables(
    vector<NodeDP>& dp_data, vector<Arc<SequencingState>*>& dp_arcs)
{
    int total_nodes = graph_->get_node_count();
    dp_data.resize(total_nodes);
    dp_arcs.assign(total_nodes, nullptr);
    Node<SequencingState>* root_node = graph_->structure.front().front();
    NodeDP& root_dp = dp_data[root_node->get_id()];
    root_dp.cost                = 0.0;
    // Nothing is scheduled yet and the only possible previous job is the depot.
    root_dp.previous_jobs       = uint64_t{1} << 0;
    root_dp.seen_jobs           = 0;
    root_dp.scheduled_weight_hi = 0;
    root_dp.scheduled_weight_lo = 0;
    root_dp.valid               = true;
}

int SequencingPathSolver::best_setup(uint64_t previous_jobs, int job_id, bool charge_most) const {
    // Dearest setup among the possible previous jobs when charging the most, the
    // cheapest one otherwise.
    int best = -1;
    uint64_t remaining = previous_jobs;
    while (remaining) {
        int index = __builtin_ctzll(remaining);
        int candidate = problem_->setup_times[index][job_id];
        if (best < 0)          best = candidate;
        else if (charge_most)  best = max(best, candidate);
        else                   best = min(best, candidate);
        remaining &= remaining - 1;
    }
    return best;
}

SequencingPathSolver::NodeDP SequencingPathSolver::describe_node(
    const Node<SequencingState>* node, const vector<NodeDP>& dp_data, int total_weight) const
{
    // Merge into one description everything the incoming paths can look like.
    NodeDP current;
    current.scheduled_weight_lo = numeric_limits<int>::max();
    for (Arc<SequencingState>* arc : node->in_arcs) {
        const NodeDP& parent = dp_data[arc->parent_node->get_id()];
        if (!parent.valid) continue;

        const int job_id = arc->variable_value;
        const uint64_t job_bit = uint64_t{1} << job_id;
        const int job_weight = problem_->weights[job_id];

        current.valid          = true;
        current.previous_jobs |= uint64_t{1} << (job_id + 1);
        current.seen_jobs     |= parent.seen_jobs | job_bit;
        // A job already seen adds no weight, and the total is capped at W_total.
        current.scheduled_weight_hi = max(current.scheduled_weight_hi,
                                          min(total_weight, parent.scheduled_weight_hi + job_weight));
        current.scheduled_weight_lo = min(current.scheduled_weight_lo,
                                          parent.scheduled_weight_lo
                                              + ((parent.seen_jobs & job_bit) ? 0 : job_weight));
    }
    return current;
}

void SequencingPathSolver::forward_pass(
    vector<NodeDP>& dp_data, vector<Arc<SequencingState>*>& dp_arcs, double cost_sign)
{
    int total_weight = 0;
    for (int w : problem_->weights) total_weight += w;

    const bool maximising = cost_sign < 0;
    // Relaxed may not overcharge, the rest may not undercharge; max swaps it.
    const bool charge_most = (relaxed_ == maximising);

    for (size_t layer_idx = 1; layer_idx < graph_->structure.size(); ++layer_idx) {
        for (Node<SequencingState>* node : graph_->structure[layer_idx]) {
            int node_id = node->get_id();
            NodeDP current = describe_node(node, dp_data, total_weight);
            if (!current.valid) continue;

            for (Arc<SequencingState>* arc : node->in_arcs) {
                const NodeDP& parent = dp_data[arc->parent_node->get_id()];
                if (!parent.valid) continue;

                int job_id   = arc->variable_value;
                int setup    = best_setup(parent.previous_jobs, job_id, charge_most);
                int duration = setup + problem_->processing_times[job_id];
                // duration * (W_total - already_scheduled), taking the end of the
                // range that keeps the bound.
                int scheduled = charge_most ? parent.scheduled_weight_lo
                                            : parent.scheduled_weight_hi;
                double arc_cost = static_cast<double>(duration) * (total_weight - scheduled);
                double candidate_cost = parent.cost + cost_sign * arc_cost;

                if (candidate_cost < current.cost) {
                    current.cost     = candidate_cost;
                    dp_arcs[node_id] = arc;
                }
            }

            dp_data[node_id] = current;
        }
    }
}

void SequencingPathSolver::extract_solution(
    const vector<NodeDP>& dp_data, const vector<Arc<SequencingState>*>& dp_arcs)
{
    Node<SequencingState>* terminal_node = graph_->structure.back().front();
    int terminal_id = terminal_node->get_id();
    double optimal_cost = dp_data[terminal_id].cost;
    if (objective_ == "max") optimal_cost = -optimal_cost;
    solution_.value = optimal_cost;
    backtrack_optimal_path(dp_arcs, terminal_id);
}

void SequencingPathSolver::backtrack_optimal_path(
    const vector<Arc<SequencingState>*>& dp_arcs, int terminal_id)
{
    solution_.path_arcs.clear();
    string path_string;
    int current_id = terminal_id;
    while (dp_arcs[current_id] != nullptr) {
        Arc<SequencingState>* arc = dp_arcs[current_id];
        path_string = "-> " + arc->to_string() + path_string;
        solution_.path_arcs.push_front(arc);
        current_id = arc->parent_node->get_id();
    }
    if (path_string.size() >= 2)
        path_string = path_string.substr(2);
    solution_.path_print = path_string;
}
